From d944c7cef14066a9b6fd8f2d4109186e44c09d1f Mon Sep 17 00:00:00 2001
From: Alexandr Baranov <alexandr.baranov@flant.com>
Date: Fri, 20 Dec 2024 16:32:48 +0300
Subject: [PATCH] pkg/nfs: wipe volume instead of removal

Use with care: this will probably slow down removal process
Also it's not totally safe (because of modern FS features, such as
journaling, CoW) and also doesn't consider specific storage features.

Signed-off-by: Alexandr Baranov <alexandr.baranov@flant.com>
---
 pkg/nfs/controllerserver.go | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/pkg/nfs/controllerserver.go b/pkg/nfs/controllerserver.go
index 726df875..62ccf9e4 100644
--- a/pkg/nfs/controllerserver.go
+++ b/pkg/nfs/controllerserver.go
@@ -260,8 +260,10 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
 		} else {
 			// delete subdirectory under base-dir
 			klog.V(2).Infof("removing subdirectory at %v", internalVolumePath)
-			if err = os.RemoveAll(internalVolumePath); err != nil {
-				return nil, status.Errorf(codes.Internal, "delete subdirectory(%s) failed with %v", internalVolumePath, err.Error())
+			wipeCmd := exec.Command("find", internalVolumePath, "-type", "f", "-exec", "shred", "-uvz", "{}", "\\;", "sync");
+			stdout, err := wipeCmd.CombinedOutput()
+			if err != nil {
+			    return nil, status.Errorf(codes.Internal, "delete subdirectory(%s) failed with %v", internalVolumePath, err.Error())
 			}
 		}
 	} else {
-- 
2.43.0

